File: PythonDesignPatterns/factory.py
"""
                 +----------------------+
                 |      Client          |
                 +----------------------+
                          |
                          | uses
                          v
                 +----------------------+
                 |      Factory         |
                 +----------------------+
                 | + createProduct()    |
                 +----------------------+
                          |
                          | creates
                          v
        -----------------------------------------
        |                 |                     |
        v                 v                     v
+----------------+ +----------------+ +----------------+
| ConcreteProduct | | ConcreteProduct | | ConcreteProduct |
|       A         | |       B         | |       C         |
+----------------+ +----------------+ +----------------+
          \               |                 /
           \              |                /
            \             |               /
             v            v              v
              +----------------------+
              |    Product (interface)|
              +----------------------+
              | + operation()        |
              +----------------------+
"""


print(__doc__)


from abc import ABC, abstractmethod

class Notification(ABC):

    @abstractmethod
    def sent():
        pass




class EmailService(Notification):

    def sent(self):
        print("Email sent....")    



class SmsService(Notification):

    def sent(self):
        print("SMS sent....")   




class PushService(Notification):

    def sent(self):
        print("Push sent....")   



#create Factory


class NotificationFactory:

    @staticmethod
    def create(type):
        if type=="email":
            return EmailService()
        if type=="Push":
            return PushService()
        if type=="SMS":
            return SmsService()



client1=NotificationFactory.create("email")
client1.sent()


client2=NotificationFactory.create("Push")
client2.sent()

File: NodeDesignPatterns/factory.ts
let doc=`
                 +----------------------+
                 |      Client          |
                 +----------------------+
                          |
                          | uses
                          v
                 +----------------------+
                 |      Factory         |
                 +----------------------+
                 | + createProduct()    |
                 +----------------------+
                          |
                          | creates
                          v
        -----------------------------------------
        |                 |                     |
        v                 v                     v
+----------------+ +----------------+ +----------------+
| ConcreteProduct | | ConcreteProduct | | ConcreteProduct |
|       A         | |       B         | |       C         |
+----------------+ +----------------+ +----------------+
          \               |                 /
           \              |                /
            \             |               /
             v            v              v
              +----------------------+
              |    Product (interface)|
              +----------------------+
              | + operation()        |
              +----------------------+
`

console.log(doc)

// class Notification{
//     sent(){
//         throw new Error("Sent function must be implemented")

//         }  
//      }


// class Email extends Notification{

// }

// a=new Email()





//interface using typescript



interface NotificationService{
    sent(): void;
}



class Email implements NotificationService{

    sent():void{
        console.log("email sent ....");
    }

}




class Push implements NotificationService{

    sent():void{
        console.log("Push sent ....")
    }

}




class SMS implements NotificationService{

    sent():void{
        console.log("SMS sent ....")
    }

}


class NotificationFactory{
    static create(type: string): NotificationService {

        if(type=="Email"){

            return new Email();
        }
        else if(type=="Push"){
            return new Push()

        }
        else if(type=="SMS"){

            return new SMS()

        }
        throw new Error("Invalid notification type");
    }
}




let client1= NotificationFactory.create("Email")
client1.sent()

client1= NotificationFactory.create("Push")
client1.sent()